### ISSUE TITLE
What triggers the messages?

### ISSUE THREAD

**Issue by Paul-Reed** [shady2k/node-red-contrib-simple-message-queue#4]
> ....and uses a feedback from a following action to release the next message

I can't see how the next message in the queue is released from the example flow you've prepared.

Is the idea to use for example; the 'return code' from an exec node feeding into a change node (to set msg.trigger) and then feed that back into your node, to release the next message?

![simple](https://user-images.githubusercontent.com/973580/46574454-944d2e00-c99b-11e8-847d-96ecdc8694e4.PNG)

**Comment by shady2k** [comment]
You are right, the description is not quite correct. Yes, example is correct. You can use feedback from your exec node. Just set msg.trigger property and feed that back into queue node.
Fixed readme in 892d5404d8b6c30f4416699d1df13805053278dd

**Comment by Paul-Reed** [comment]
Hhmm, the example isn't working because the first message isn't being passed through your node to the exec node to be fed back to trigger msg2.
I thought that if the queue wasn't busy, the msg should pass straight through, or have I got this wrong?

**Comment by shady2k** [comment]
You can achieve this behavior if set "Bypass first message?" flag to True in node settings.
Queue never busy, because it's purpose to store messages only. What it's role, if it will pass through every message?
In my project, I'm use external event to trigger message. If you need other behavior, just describe it in detail.

**Comment by Paul-Reed** [comment]
Let me share a example flow to demonstrate the issue I'm facing.

If you click 'test' once, the msg gets passed through, click it again after say 10 seconds and it doesn't, it just stays in the queue.
```
[
    {
        "id": "f9dcfad7.579938",
        "type": "inject",
        "z": "c53060.842a0fa",
        "name": "",
        "topic": "",
        "payload": "test",
        "payloadType": "str",
        "repeat": "",
        "crontab": "",
        "once": false,
        "x": 210,
        "y": 770,
        "wires": [
            [
                "6accc9a8.ff82c8"
            ]
        ]
    },
    {
        "id": "702ad25.1555d2c",
        "type": "debug",
        "z": "c53060.842a0fa",
        "name": "",
        "active": true,
        "console": "false",
        "complete": "true",
        "x": 990,
        "y": 790,
        "wires": []
    },
    {
        "id": "73937bdb.634104",
        "type": "inject",
        "z": "c53060.842a0fa",
        "name": "",
        "topic": "",
        "payload": "",
        "payloadType": "date",
        "repeat": "",
        "crontab": "",
        "once": false,
        "x": 210,
        "y": 850,
        "wires": [
            [
                "4443adbc.cd0714"
            ]
        ]
    },
    {
        "id": "4443adbc.cd0714",
        "type": "change",
        "z": "c53060.842a0fa",
        "name": "",
        "rules": [
            {
                "t": "set",
                "p": "trigger",
                "pt": "msg",
                "to": "1",
                "tot": "str"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 400,
        "y": 890,
        "wires": [
            [
                "6accc9a8.ff82c8",
                "ed698804.410708"
            ]
        ]
    },
    {
        "id": "f2ff7d81.21d18",
        "type": "inject",
        "z": "c53060.842a0fa",
        "name": "",
        "topic": "",
        "payload": "",
        "payloadType": "date",
        "repeat": "",
        "crontab": "",
        "once": false,
        "x": 210,
        "y": 810,
        "wires": [
            [
                "142ccad9.a6edb5"
            ]
        ]
    },
    {
        "id": "142ccad9.a6edb5",
        "type": "change",
        "z": "c53060.842a0fa",
        "name": "",
        "rules": [
            {
                "t": "set",
                "p": "reset",
                "pt": "msg",
                "to": "1",
                "tot": "str"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 410,
        "y": 850,
        "wires": [
            [
                "6accc9a8.ff82c8"
            ]
        ]
    },
    {
        "id": "6accc9a8.ff82c8",
        "type": "simple-queue",
        "z": "c53060.842a0fa",
        "name": "queue1",
        "firstMessageBypass": true,
        "bypassInterval": "0",
        "x": 640,
        "y": 840,
        "wires": [
            [
                "702ad25.1555d2c",
                "ee2a8e97.a8704"
            ]
        ]
    },
    {
        "id": "bd12e611.3b9248",
        "type": "function",
        "z": "c53060.842a0fa",
        "name": "",
        "func": "node.send({trigger:\"1\"});",
        "outputs": 1,
        "noerr": 0,
        "x": 820,
        "y": 750,
        "wires": [
            [
                "f885b0ae.14c7b",
                "6accc9a8.ff82c8"
            ]
        ]
    },
    {
        "id": "f885b0ae.14c7b",
        "type": "debug",
        "z": "c53060.842a0fa",
        "name": "",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "true",
        "x": 990,
        "y": 750,
        "wires": []
    },
    {
        "id": "ee2a8e97.a8704",
        "type": "delay",
        "z": "c53060.842a0fa",
        "name": "",
        "pauseType": "delay",
        "timeout": "3",
        "timeoutUnits": "seconds",
        "rate": "1",
        "nbRateUnits": "1",
        "rateUnits": "second",
        "randomFirst": "1",
        "randomLast": "5",
        "randomUnits": "seconds",
        "drop": false,
        "x": 660,
        "y": 750,
        "wires": [
            [
                "bd12e611.3b9248"
            ]
        ]
    },
    {
        "id": "ed698804.410708",
        "type": "debug",
        "z": "c53060.842a0fa",
        "name": "",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "true",
        "x": 990,
        "y": 830,
        "wires": []
    }
]
```

**Comment by shady2k** [comment]
That's right. Do you need reset first message flag after configurable specific timeout?

**Comment by shady2k** [comment]
But every message - it's a new message. They will be pass through always.

**Comment by Paul-Reed** [comment]
But in a continuing sequence of msg's, when would you send a reset?
I don't see how this node can work in its present format, because in Dave CJ's original flow, if the queue was empty, the msg was automatically bypassed, and passed straight through.
In your node, that doesn't appear to happen, and relies upon a manually set bypass.

If you are happy this is working as intended, can you share a flow where it does work, and deals with a continuing flow of msg's, being inputted.

**Comment by Paul-Reed** [comment]
...this is the code which Dave CJ uses to pass the msg through if queue is not busy.
```
else {
    if (context.busy) {
        // if busy add to queue
        context.queue.push(msg.payload);
    }
    else {
        // otherwise we are empty so just pass through and set busy flag
        context.busy = true;
        return msg;
    }
}
```

**Comment by shady2k** [comment]
Ok, check d273c396734481cc5b907de5073cc8aca54fa069

**Comment by Paul-Reed** [comment]
Just done a quick check, and yes that does work.

I'm short of time at the moment, but I'll implement the node into a real flow this evening, which will give it a good test.

I'll reply back this evening.

Thanks for the speedy update!

**Comment by shady2k** [comment]
Thanks, I will be waiting for your reply.

**Comment by Paul-Reed** [comment]
Works fine. I've added it to two of my flows, and it's now correctly releasing messages.
Have tested by injecting test data to the node, at random millisecond intervals, and monitoring the queue, incoming & outgoing, and it's behaved impeccably.

Thanks for sorting, I've just rated the node on node_RED with 5 stars.

Paul

![ring](https://user-images.githubusercontent.com/973580/46586870-6a1f6d00-ca7c-11e8-9f1a-ed725e031185.PNG)
